### DISCUSSION TITLE
Visualizing models with hiddenlayer and torchviz

### DISCUSSION THREAD

**Discussion by hoseinhamzei** [ultralytics/yolov5#8115] [Q&A] [answered]
Hello I'm trying to visualize the yolov5 models with the hiddenlayer and torchviz libraries so that i can compare and customize models quickly without having to train them and without tensorboard. but I don't know what is the exact input size of yolov5 models to pass to them I tried the input size of the first layer of yolov5s ([32, 3, 6, 6]) but it throws an error:
 `Sizes of tensors must match except in dimension 1. Expected size 2 but got size 1 for tensor number 1 in the list.`

in the below code I don't know what to pass as the second argument of hidden layer and the first argument of torchviz

```
import os
import sys
from pathlib import Path
import torch
import hiddenlayer as hl
from torchviz import make_dot

FILE = Path(__file__).resolve()
ROOT = FILE.parents[0]  # YOLOv5 root directory
if str(ROOT) not in sys.path:
    sys.path.append(str(ROOT))  # add ROOT to PATH
ROOT = Path(os.path.relpath(ROOT, Path.cwd()))  # relative

from models.yolo import Model


LOCAL_RANK = int(os.getenv('LOCAL_RANK', -1))  # https://pytorch.org/docs/stable/elastic/run.html
RANK = int(os.getenv('RANK', -1))
WORLD_SIZE = int(os.getenv('WORLD_SIZE', 1))

model = Model(cfg="models/yolov5s.yaml")

#print("model:", model)
# hidden layer
hl.build_graph(model, torch.zeros([32, 3, 6, 6]))

# torchviz
x = torch.zeros(32, 3, 6, 6).requires_grad_(True)
y = model(x)

make_dot(y, params=dict(list(model.named_parameters()))).render("torchviz", format="png")

```

 the current code return an error:
 `Sizes of tensors must match except in dimension 1. Expected size 2 but got size 1 for tensor number 1 in the list.`

would appreciate any help

https://github.com/waleedka/hiddenlayer
https://github.com/szagoruyko/pytorchviz

**Comment by zhiqwang** [comment]
Just FYI @hoseinhamzei , I've done some similar work at https://zhiqwang.com/yolov5-rt-stack/notebooks/model-graph-visualization.html before

<a href="https://raw.githubusercontent.com/zhiqwang/yolov5-rt-stack/main/notebooks/assets/yolov5_graph_visualize.svg"><img src="https://raw.githubusercontent.com/zhiqwang/yolov5-rt-stack/main/notebooks/assets/yolov5_graph_visualize.svg" alt="YOLO model visualize" width="390"/></a>

**Comment by glenn-jocher** [comment, marked as answer]
@hoseinhamzei 👋 Hello! Thanks for asking about YOLOv5 🚀 **architecture visualization**. we've made visualizing YOLOv5 🚀 architectures super easy. There are 3 main ways:

## `model.yaml`

Each model has a corresponding yaml file that displays the model architecture. Here is YOLOv5s, defined by `yolov5s.yaml`:
https://github.com/ultralytics/yolov5/blob/1a3ecb8b386115fd22129eaf0760157b161efac7/models/yolov5s.yaml#L12-L48

## TensorBoard Graph

Simply start training a model, and then view the TensorBoard Graph for an interactive view of the model architecture. This example shows YOLOv5s viewed in our [Notebook](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) –  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

```python
# Tensorboard
%load_ext tensorboard
%tensorboard --logdir runs/train

# Train YOLOv5s on COCO128 for 3 epochs
python train.py --weights yolov5s.pt --epochs 3
```

<img width="893" alt="Screenshot 2021-04-11 at 01 10 09" src="https://user-images.githubusercontent.com/26833433/114286928-349bd600-9a63-11eb-941f-7139ee6cd602.png">

## Netron viewer

Use [https://netron.app](https://netron.app/) to view exported ONNX models:

```
python export.py --weights yolov5s.pt --include onnx --simplify
```

<img width="890" alt="Screen Shot 2022-04-29 at 11 09 23 AM" src="https://user-images.githubusercontent.com/26833433/165999628-0d6095ce-58dd-49b7-a04e-25d0995e5493.png">


Good luck 🍀 and let us know if you have any other questions!

**Reply by hoseinhamzei** [reply]
thanks a lot exporting model to onnx and using it in netron showed more detail and I've tried tensorboard and it's even better however what I need to do is visualizing the model graph without weights and just with yaml files because I'm working on a modified model yaml file, if that isn't possible is there a way or a function or something to validate a yaml file to find out if the model is valid and will work?

**Reply by glenn-jocher** [reply]
@hoseinhamzei yes of course:
```
python models/yolo.py --cfg model.yaml
```
